src/tools/cephfs/cephfs-shell

```diff
@@ -779,6 +779,134 @@ class CephFSShell(Cmd):
+            # sometimes, cwd is like /dir1/dir2, and there is a source file
+            # foo then path would be like /dir1/dir2foo which is not expected
+            # therefore checking existence of `/` here.
+            if cwd[len(cwd) - 1] != '/':
+                cwd = cwd + '/'
+            cwd = cwd.encode('utf-8')
+            if target_last_char_slash is True:
+                if is_dir_exists(args.target):
+                    pass
+                else:
+                    set_exit_code_msg(errno.ENOTDIR,
+                                      f"ln: failed to access "
+                                      f"'{args.target.decode('utf-8')}': "
+                                      f"Not a directory")
+                    return
+            file_or_dir = os.path.basename(os.path.normpath(args.target))
+            link_name = cwd + file_or_dir
+
+            if (is_file_exists(link_name) or is_dir_exists(
+                    link_name)) and not args.force:
+                set_exit_code_msg(errno.ENOENT,
+                                  msg=f"ln: failed to create link "
+                                      f"{link_name.decode('utf-8')}: "
+                                      f"exists")
+                return
+        else:
+            if is_dir_exists(args.link_name):
+                dest = args.link_name.decode('utf-8').rstrip('/')
+                dest_first_half = dest.encode('utf-8') + b'/'
+                if is_file_exists(args.target):
+                    if target_last_char_slash is True:
+                        set_exit_code_msg(errno.ENOTDIR,
+                                          "ln: failed to access "
+                                          f"'{args.target.decode('utf-8')}': "
+                                          "Not a directory")
+                        return
+                    dest_file = os.path.basename(os.path.normpath(args.target))
+                    link_name = dest_first_half + dest_file
+
+                elif is_dir_exists(args.target):
+                    dest_dir = os.path.basename(os.path.normpath(args.target))
+                    link_name = dest_first_half + dest_dir
+
+            else:
+                # if the destination is not a file or a dir then:
+                # accept it as file so the end part of path cannot have
+                #    a `/` succeeding it.
+                test_path = args.link_name.decode('utf-8')
+                if test_path[len(test_path) - 1] == '/':
+                    set_exit_code_msg(errno.ENOENT, f"'{test_path}': "
+                                                    f"No such file or "
+                                                    f"directory")
+                    return
+                else:
+                    link_name = test_path.encode('utf-8')
+
+        if args.force:
+            try:
+                cephfs.lstat(os.path.join(b'', link_name))
+            except libcephfs.Error:
```

Comment:
Isn't libcephfs.Error too broad? We don't want ignore all errors. Can we use something more specific?

Reply:
The change is not incorporated yet.

Reply:
I think we decided to keep it the way it is right?

Reply:
`ObjectNotFound ` will be more appropriatre. `libcephfs.Error` is too broad. We don't want to ignore all other exceptions. Doing so will amount to adding a bug.

I've confirmed this as well.

I added following line of code in to on top of your patch -
```
x = cephfs.lstat(os.path.join(b'', b'dir111'))
```
And this is what I get -
```
-> x = cephfs.lstat(os.path.join(b'', b'dir111'))
(Pdb) n
cephfs.ObjectNotFound: error in stat: dir111: No such file or directory [Errno 2]
```

Reply:
After this the comment in next line ("# if file doesn't exist, unlink won't happen") will be redundant.

Reply:
pushed the changes